haskell/src/Stablecoin/Client/Contract.hs

```diff
@@ -2,43 +2,46 @@
 -- SPDX-License-Identifier: MIT
 
 module Stablecoin.Client.Contract
-  ( InitialStorageData(..)
+  ( InitialStorageOptions(..)
+  , InitialStorageData(..)
   , mkInitialStorage
   ) where
 
 import Lorentz as L
 import Lorentz.Contracts.Spec.TZIP16Interface (MetadataMap)
 import Lorentz.Contracts.Stablecoin (Expiry, Roles(..), Storage(..))
-import Morley.Tezos.Address.Alias (AddressOrAlias)
+import Morley.Tezos.Address (ContractAddress, L1Address)
+import Stablecoin.Client.L1AddressOrAlias (ContractAddressOrAlias, L1AddressOrAlias)
 import Stablecoin.Client.Parser (ContractMetadataOptions(..))
 
-type family ComputeContractMetadataStorageType a where
-  ComputeContractMetadataStorageType Address = MetadataMap
-  ComputeContractMetadataStorageType AddressOrAlias = ContractMetadataOptions
+-- | The options needed to create the data for the contract's initial storage.
+data InitialStorageOptions = InitialStorageOptions
+  { isoMasterMinter :: L1AddressOrAlias
+  , isoContractOwner :: L1AddressOrAlias
+  , isoPauser :: L1AddressOrAlias
+  , isoTransferlist :: Maybe ContractAddressOrAlias
+  , isoTokenSymbol :: Text
+  , isoTokenName :: Text
+  , isoTokenDecimals :: Natural
+  , isoContractMetadataStorage :: ContractMetadataOptions
+  , isoDefaultExpiry :: Expiry
+  }
```

Comment:
I have... some thoughts on DRY :sweat_smile: 

I see it as a means to an end; not as an end itself.

IMHO, DRY is not strictly about reducing the SLOC count.
The point of DRY is to avoid situations where you have the same (business, domain, whatever) logic duplicated across the codebase, the requirements change, and now you have to update multiple places. If you update some but forget others, you now have a bug.

In another sense, DRY is also useful when designing a public API: you don't want to have multiple ways of doing the same thing, that makes it way harder to use and remember. Homogeneity and consistency are great.

IMO, neither of these 2 concerns apply in this situation.

--- 

More importantly, these 2 data types are not _conceptually_ the same thing.
They're only _accidentally_ similar.

`InitialStorageOptions` is where we store the data we need to pass to the `deploy` command, and `InitialStorageData` is the where we store the data to create the `Storage`.

They _happen_ to have the same number of fields, with similar types, but they don't necessarily have to.
It's possible `deploy` could do some work to "crunch" 3 fields from `InitialStorageOptions` into a single field.

--- 

The problem with creating abstractions to encapsulate _accidental_ similarities is that, later, when we realize they should be 2 separate concepts, developers will be deterred from deleting the abstraction.

Going with the example from above: Say we create an abstraction `InitialStorageInfo stage`. If, later, we need to change `deploy` to crunch 3 fields into a single field, the developer will face 2 options:
1. Split `InitialStorageInfo stage` into two distinct data types, one with `n` fields and the other with `n+2` fields.
2. Move the crunching logic somewhere else, possibly to `mkInitialStorage`?

I think most devs would be inclined to go with (2) to avoid deleting the abstraction, even when (1) would be the best solution.


----


There are even more lax schools of thought like the [Rule of three](https://en.m.wikipedia.org/wiki/Rule_of_three_(computer_programming)), which I'm a bit on the fence about.

I do see the point of the rule, to avoid premature misguided abstractions, but I also think you can do that _without_ waiting for the pattern to emerge thrice.

Reply:
That being said, I just realized `InitialStorageData` and `mkInitialStorage` are only used in exactly one place, so I just inlined them :rofl:

Reply:
Okay, all good points, and inlining is a better solution anyway :+1: 

That said, I could argue this similarity is not exactly accidental, but that's beside the point now, so I won't :laughing: